### PR TITLE
#184: ShadedRelief implementation

### PR DESCRIPTION
**Pull request by etj**

### CODE REVIEW

File: jt-shadedrelief/pom.xml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    <url>http://maven.apache.org</url>
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>2.3.2</version>
+                <configuration>
+                    <source>1.8</source>
+                    <target>1.8</target>
```
[the remark this answers is not in the document]
Comment (etj) [reply jt-shadedrelief/pom.xml]: There are some minor constructs only available for java8.
Should this module be java7 compatible?

File: jt-shadedrelief/src/test/java/it/geosolutions/jaiext/shadedrelief/ShadedReliefTest.java [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+
+        for (Integer i : typesToTest) {
+            RenderedOp shaded = ShadedReliefDescriptor.create(testImages[i], roi, noDataD, destNoData, resx, resy, vex, ves, alt, az, algorithm, null);
+            check(i, testImages[i], shaded, roiUsed, roi, nodataUsed, noData[i]);
+            shaded.dispose();
+        }
+    }
+
+    private void check(int type, RenderedImage src, RenderedOp shaded, boolean roiUsed,
+           ROI roi, boolean nodataUsed, Range noData) {
```
[the remark this answers is not in the document]
Comment (etj) [reply jt-shadedrelief/src/test/java/it/geosolutions/jaiext/shadedrelief/ShadedReliefTest.java]: There is some confusion between source nodata and destination nodata indeed.

File: jt-shadedrelief/src/main/java/it/geosolutions/jaiext/shadedrelief/ShadedReliefDescriptor.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    private static final String[][] resources = {
+        {"GlobalName", "ShadedRelief"},
+        {"LocalName", "ShadedRelief"},
+        {"Vendor", "it.geosolutions.jaiext"},
+        {"Description", "desc"},
+        {"Version", "ver"},
+        {"arg0Desc", "Region of interest"},
+        {"arg1Desc", "Input NoData"},
+        {"arg2Desc", "Destination NoData"},
+        {"arg3Desc", "X resolution"},
```
Comment (aaime) [inline jt-shadedrelief/src/main/java/it/geosolutions/jaiext/shadedrelief/ShadedReliefDescriptor.java]: Need some better description of the following params (here, or at least, somewhere):
* x resolution and y resolution (and I have a hunch of what they do, but best not having to guess)
* vertical exxageration
* the unit scale ratio (believe it's for data in degrees, again, need some explanation)
* altitude, azimuth (of the sun? in what unit? degrees, radians?)

File: jt-shadedrelief/src/main/java/it/geosolutions/jaiext/shadedrelief/ShadedReliefDescriptor.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+     *
+     * @return An array of property generators.
+     */
+    public PropertyGenerator[] getPropertyGenerators() {
+        PropertyGenerator[] pg = new PropertyGenerator[1];
+        pg[0] = new AreaOpPropertyGenerator();
+        return pg;
+    }
+
+    public static RenderedOp create(
```
Comment (aaime) [inline jt-shadedrelief/src/main/java/it/geosolutions/jaiext/shadedrelief/ShadedReliefDescriptor.java]: Same as above, need javadoc with descriptions, or pointers to a description

**Comment by etj** [comment]
@aaime @dromagnoli many fixes and improvements here and there.

You can also run the algorithm on command line, using a comand like this:

```
mvn exec:java -Dexec.classpathScope=test  -Dexec.mainClass=it.geosolutions.jaiext.shadedrelief.cli.ShadedReliefCommand  -Dexec.args="-f srtm_boulder.tiff -o boulder_shaded.tif -z 1 -alt 64 -s 111120  -az 126  --alg ZT -x 0.000277 -y 0.000277" 
```

The valid params are:
```
usage: shadedrelief.sh [-alg <algorithm>] [-alt <val>] [-az <val>] [-dn
       <val>] -f <file> -o <file> [-s <val>] [-sn <val>] [-x <val>] [-y
       <val>] [-z <val>]
hillshading
 -alg,--algorithm <algorithm>   Algorithm. May be ZT, ZTC, C
 -alt,--alt <val>               Sun altitude
 -az,--azimuth <val>            Sun azimuth
 -dn,--dstnodata <val>          Value of nodata in out file
 -f,--in <file>                 Input file
 -o,--out <file>                Output file
 -s,--scale <val>               Elevation unit to 2D unit scale ratio
 -sn,--srcnodata <val>          Value of nodata in source file
 -x,--resx <val>                X resolution
 -y,--resy <val>                Y resolution
 -z,--zeta <val>                Vertical relief factor (exaggeration)
```

**Comment by dromagnoli** [comment]
Note that the x,y resolutions are the one of the underlying raster geotransformation.
So resy is usually a negative number.
With this in mind, the results look basically identical

**ZevenbergenThorne algorithm (left side: jai-ext; right side: gdaldem)**
shadedreliefcompare


**ZevenbergenThorne Combined algorithm (left side: jai-ext; right side: gdaldem)**
shadedreliefcomparecombined

**Pull request merged**